File: Licenta.Server/Services/ProjectService.cs
using Licenta.Server.DataLayer.Dto;
using Licenta.Server.DataLayer.Models;
using Licenta.Server.DataLayer.Utils;

namespace Licenta.Server.Services
{
    public class ProjectService
    {
        private readonly UnitOfWork _unitOfWork;

        public ProjectService(UnitOfWork unitOfWork)
        {
            _unitOfWork = unitOfWork;
        }
        public async Task<List<Project>> GetAllProjects()
        {
            return await _unitOfWork.ProjectsRepository.GetAllProjects();
        }
        public async Task<Project> GetProjectById(Guid id)
        {
            return await _unitOfWork.ProjectsRepository.GetProjectById(id);
        }
        public async Task AddProject(CreateProjectDTO project, Guid ownerId)
        {
            await _unitOfWork.ProjectsRepository.AddProject(project, ownerId);
        }
        public async Task<Project?> DeleteProject(Guid id)
        {
            return await _unitOfWork.ProjectsRepository.DeleteProject(id);
        }
        public async Task UpdateProject(AddProjectDto project)
        {
            await _unitOfWork.ProjectsRepository.UpdateProject(project);
        }
        public async Task<List<Project>> GetAllProjectsForOwner(string email)
        {
            return await _unitOfWork.ProjectsRepository.GetAllProjectsForOwner(email);
        }
        public async Task<List<Issue>> GetAllIssuesForProject(Guid id)
        {
            return await _unitOfWork.ProjectsRepository.GetAllIssuesForProject(id);
        }
        public async Task<List<Sprint>> GetAllSprintsForProject(Guid id)
        {
            return await _unitOfWork.ProjectsRepository.GetAllSprintsForProject(id);
        }
        public async Task<List<User>> GetAllUsersForProject(Guid id)
        {
            return await _unitOfWork.ProjectsRepository.GetAllUsersForProject(id);
        }
        public async Task AddUserToProject(Guid projectId, string email)
        {
            await _unitOfWork.ProjectsRepository.AddMemberToProject(email, projectId);
        }
        public async Task RemoveUserFromProject(Guid projectId, string email)
        {
            await _unitOfWork.ProjectsRepository.RemoveMemberFromProject(email, projectId);
        }
        public async Task<List<Project>> GetAllProjectsForMemeber(string email)
        {
            return await _unitOfWork.ProjectsRepository.GetAllProjectsForMemeber(email);
        }
        public async Task AddNewIssueStatusToProject(Guid projectId, IssueStatus status)
        {
            await _unitOfWork.ProjectsRepository.AddNewIssueStatusToProject(projectId, status);
        }
        public async Task AddIssueStatusToProject(Guid projectId, int issueStatusId)
        {
            await _unitOfWork.ProjectsRepository.AddIssueStatusToProject(projectId, issueStatusId);
        }
        public async Task<List<IssueStatus>> GetIssueStatusFromProject(Guid projectId)
        {
            return await _unitOfWork.ProjectsRepository.GetIssueStatusFromProject(projectId);
        }
        public async Task RemoveIssueStatusFromProject(Guid projectId, int issueStatusId)
        {
            await _unitOfWork.ProjectsRepository.RemoveIssueStatusFromProject(projectId, issueStatusId);
        }
        public async Task<List<User>> GetAllAdministratorsForProject(Guid projectId)
        {
            return await _unitOfWork.ProjectsRepository.GetAllAdministratorsForProject(projectId);
        }
        public async Task AddAdministratorToProject(Guid projectId, string email)
        {
            await _unitOfWork.ProjectsRepository.AddAdministratorToProject(projectId, email);
        }
        public async Task RemoveAdministratorFromProject(Guid projectId, string email)
        {
            await _unitOfWork.ProjectsRepository.RemoveAdministratorFromProject(projectId, email);
        }
        public async Task RemoveSprintFromProject(Guid projectId, Guid sprintId)
        {
            await _unitOfWork.ProjectsRepository.RemoveSprintFromProject(projectId, sprintId);
        }
        public async Task AddIssueLabel(Guid projectId, IssueLabel label)
        {
            await _unitOfWork.ProjectsRepository.AddIssueLabel(projectId, label);
        }
        public async Task  RemoveIssueLabel(Guid projectId, int labelId)
        {
            await _unitOfWork.ProjectsRepository.RemoveIssueLabel(projectId, labelId);
        }
        public async Task<List<IssueLabel>> GetIssueLabelsFromProject(Guid projectId)
        {
            return await _unitOfWork.ProjectsRepository.GetAllLabelsForProject(projectId);
        }
        public async Task RemoveIssueFromProject(Guid projectId, Guid issueId)
        {
            await _unitOfWork.ProjectsRepository.RemoveIssueFromProject(projectId, issueId);
        }
    } 
}

File: Licenta.Frontend/src/app/Services/project-service.ts
import { Injectable } from '@angular/core';
import { User } from '../Models/User';
import { UserRegister } from '../Models/UserRegister';
import { HttpClient } from '@angular/common/http';
import { map } from 'rxjs/operators';
import { Skill } from '../Models/Skill';
import { Project } from '../Models/Project';
import { AddProjectDto } from '../Models/DTO/AddProjectDto';
import { Observable } from 'rxjs';
import { Sprint } from '../Models/Sprint';
import { IssueLabel } from '../Models/IssueLabel';
import { PriorityLevel } from '../Models/PriorityLevel';
@Injectable({
  providedIn: 'root',
})
export class ProjectService {
  public projectData: Project | null = null;
  public userData: User | null = null;
  public selectedProject: Project | null = null;
  private apiUrl = 'https://localhost:7261';
  private projectApiUrl = `${this.apiUrl}/api/Project`;
  private userApiUrl = `${this.apiUrl}/User`;

  constructor(private http: HttpClient) {}

  setUserProject(project: Project) {
    localStorage.setItem('project', JSON.stringify(project));
  }

  getUserDataFromLocalStorage(): User | null {
    const user = localStorage.getItem('user');
    return user ? JSON.parse(user) : null;
  }

  getUserProjectsFromApi(email: string) {
    const url = `${this.projectApiUrl}/GetAllProjectsForMemeber?email=${encodeURIComponent(email)}`;
    return this.http.get<Project[]>(url).pipe(
      map((response: any) => {
        return response as Project[];
      })
    );
  }

  setSelectedProject(project: Project) {
    this.projectData = project;
    console.log(this.projectData);
    localStorage.setItem('selectedProject', JSON.stringify(project));
  }

  addProject(project: AddProjectDto, ownerId?: number): Observable<any> {
    const url = `${this.projectApiUrl}/AddProject?ownerId=${ownerId}`;
    return this.http.post(url, project).pipe(
      map((response) => {
        return response;
      })
    );
  }

  deleteProject(projectId: string): Observable<any> {
    const url = `${this.projectApiUrl}/DeleteProject?id=${projectId}`;
    return this.http.delete(url).pipe(
      map((response) => {
        return response;
      })
    );
  }

  leaveProject(projectId: string): Observable<any> {
    const url = `${this.projectApiUrl}/RemoveUserFromProject/${projectId}${this.userData?.email}`;
    return this.http.delete(url).pipe(
      map((response) => {
        return response;
      })
    );
  }

  getSelectedProject(): Project | null {
    if (this.selectedProject === null) {
      this.selectedProject = JSON.parse(
        localStorage.getItem('selectedProject') || 'null'
      );
    }
    return this.selectedProject;
  }

  joinProject(key: string): Observable<any> {
    var user = this.getUserDataFromLocalStorage();
    const url = `${this.userApiUrl}/JoinProject?email=${user?.email}&key=${key}`;
    return this.http.post(url, null).pipe(
      map((response) => {
        return response;
      })
    );
  }

  getAllMembers(): Observable<User[]> {
    var projectId = this.getSelectedProject()?.projectId;
    const url = `${this.projectApiUrl}/GetAllUsersForProject?id=${projectId}`;
    return this.http.get<User[]>(url).pipe(
      map((response) => {
        return response as User[];
      })
    );
  }

  getAllSprintsForProject(): Observable<Sprint[]> {
    const projectId = this.getSelectedProject()?.projectId;
    const url = `${this.projectApiUrl}/GetAllSprintsForProject?id=${projectId}`;
    return this.http.get<Sprint[]>(url).pipe(
      map((response) => {
        return response as Sprint[];
      })
    );
  }

  getAllAdministratorsForProject(): Observable<User[]> {
    const projectId = this.getSelectedProject()?.projectId;
    const url = `${this.projectApiUrl}/GetAllAdministratorsForProject?projectId=${projectId}`;
    return this.http.get<User[]>(url).pipe(
      map((response) => {
        return response as User[];
      })
    );
  }
  addAdminstratorToProject(email: string): Observable<any> {
    const projectId = this.getSelectedProject()?.projectId;
    const url = `${this.projectApiUrl}/AddAdministratorToProject?projectId=${projectId}&email=${email}`;
    return this.http.post(url, null).pipe(
      map((response) => {
        return response;
      })
    );
  }

  removeMemberFromProject(email:string): Observable<any> {
    const projectId = this.getSelectedProject()?.projectId;
    const url = `${this.projectApiUrl}/RemoveUserFromProject?projectId=${projectId}&email=${email}`;
    return this.http.put(url, null).pipe(
      map((response) => {
        return response;
      })
    );
  }
  getAllLabelsForProject(): Observable<IssueLabel[]> {
    const projectId = this.getSelectedProject()?.projectId;
    const url = `${this.projectApiUrl}/getLabelsForProject?projectId=${projectId}`;
    return this.http.get<IssueLabel[]>(url).pipe(
      map((response) => {
        return response as IssueLabel[];
      })
    );
  }
  removeIssueFromProject(issueId: string): Observable<any> {
    const projectId = this.getSelectedProject()?.projectId;
    const url = `${this.projectApiUrl}/removeIssueFromProject?projectId=${projectId}&issueId=${issueId}`;
    return this.http.delete(url).pipe(
      map((response) => {
        console.log("Repsonse" + response);
        return response;
      })
    );
  }
  addLabelToProject(label: IssueLabel): Observable<any> {
    const projectId = this.getSelectedProject()?.projectId;
    const url = `${this.projectApiUrl}/addLabelToProject?projectId=${projectId}`;
    return this.http.post(url, label).pipe(
      map((response) => {
        return response;
      })
    );
  }
  removeAdminstratorFromProject(email: string): Observable<any> {
    const projectId = this.getSelectedProject()?.projectId;
    const url = `${this.projectApiUrl}/RemoveAdministratorFromProject?projectId=${projectId}&email=${email}`;
    return this.http.delete(url).pipe(
      map((response) => {
        return response;
      })
    );
  }
  getPriorities():string[]
  {
    var priorities = [
      {
        value: PriorityLevel.Low,
        name: 'Low',
        color: 'green',
        icon: 'arrow_downward',
        id:0
      },
      {
        value: PriorityLevel.Medium,
        name: 'Medium',
        color: '#DAA520',
        icon: 'arrow_right_alt',
        id:1
      },
      {
        value: PriorityLevel.High,
        name: 'High',
        color: 'orange',
        icon: 'arrow_upward',
        id:2
      },
      {
        value: PriorityLevel.Critical,
        name: 'Critical',
        color: 'red',
        icon: 'error',
        id:3
      },
    ];
    return priorities.map(x=>x.id.toString())
  }
}